### DISCUSSION TITLE
How to run the contraction under the low memory constraint?

### DISCUSSION THREAD

**Discussion by rht** [NVIDIA/cuQuantum#63] [Q&A] [answered]
According to https://docs.nvidia.com/cuda/cuquantum/cutensornet/overview.html#slicing,
> In order to fit a tensor network contraction into available device memory, as specified by workspaceSizeConstraint, it may be necessary to use slicing (also known as variable projection or bond cutting).

I'm running a QAOA circuit as described in https://arxiv.org/abs/2012.02430 (MaxCut problem for random graphs of degree 3). I have explicitly ensured that slicing is enabled, and have tweaked various slicing settings as shown in the following code snippet
```python
    optimize_options = {
        "samples": 5,  # default is 0 (disabled)
        "slicing": {
            "disable_slicing": 0,
            "memory_model": 1,  # 0 is heuristic, 1 is cutensor (default)
            "min_slices": 10,  # default is 1
            "slice_factor": 2,  # default is 32
        },
        "cost_function": 0,  # 0 for FLOPS (default), 1 for time
        "reconfiguration": {
            "num_iterations": 800,  # default is 500; good values are within 500-1000
            # Higher number means more time spent in reconfiguration, scales exponentially
            "num_leaves": 10,  # default is 8
        },
    }
```
I split the contraction step into `contract_path` and `contract`. I'm not sure whether this `optimize_options` should be passed to `contract_path` or `contract` (I did both).

At 30 qubits, on a NVIDIA A100 40 GB, I constantly got this error, which means the contraction path is not efficient enough. But even so, shouldn't the slicing ensures that the contraction can run, albeit taking longer time (as a trade off for lower memory)?

```
Traceback (most recent call last):
  File "cutn_notebook_benchmark_simplified.py", line 129, in <module>
    run_multiple_methods(ansatz, qubits, index=i)
  File "cutn_notebook_benchmark_simplified.py", line 88, in run_multiple_methods
    out = common_cutn.run_with_cutn(circuit, pauli_string)
  File "/home/cuquantum/common_cutn.py", line 82, in run_with_cutn
    output = contract(
  File "/home/cuquantum/conda/envs/cuquantum-23.03/lib/python3.8/site-packages/cuquantum/cutensornet/tensor_network.py", line 963, in contract
    opt_info = network.contract_path(optimize=optimize)
  File "/home/cuquantum/conda/envs/cuquantum-23.03/lib/python3.8/site-packages/cuquantum/cutensornet/_internal/utils.py", line 474, in inner
    result = wrapped_function(*args, **kwargs)
  File "/home/cuquantum/conda/envs/cuquantum-23.03/lib/python3.8/site-packages/cuquantum/cutensornet/_internal/utils.py", line 443, in inner
    raise e
  File "/home/cuquantum/conda/envs/cuquantum-23.03/lib/python3.8/site-packages/cuquantum/cutensornet/_internal/utils.py", line 435, in inner
    result = wrapped_function(*args, **kwargs)
  File "/home/cuquantum/conda/envs/cuquantum-23.03/lib/python3.8/site-packages/cuquantum/cutensornet/tensor_network.py", line 568, in contract_path
    self._calculate_workspace_size()
  File "/home/cuquantum/conda/envs/cuquantum-23.03/lib/python3.8/site-packages/cuquantum/cutensornet/_internal/utils.py", line 474, in inner
    result = wrapped_function(*args, **kwargs)
  File "/home/cuquantum/conda/envs/cuquantum-23.03/lib/python3.8/site-packages/cuquantum/cutensornet/_internal/utils.py", line 474, in inner
    result = wrapped_function(*args, **kwargs)
  File "/home/cuquantum/conda/envs/cuquantum-23.03/lib/python3.8/site-packages/cuquantum/cutensornet/tensor_network.py", line 384, in _calculate_workspace_size
    cutn.workspace_compute_contraction_sizes(self.handle, self.network, self.optimizer_info_ptr, self.workspace_desc)
  File "cuquantum/cutensornet/cutensornet.pyx", line 679, in cuquantum.cutensornet.cutensornet.workspace_compute_contraction_sizes
  File "cuquantum/cutensornet/cutensornet.pyx", line 696, in cuquantum.cutensornet.cutensornet.workspace_compute_contraction_sizes
  File "cuquantum/cutensornet/cutensornet.pyx", line 240, in cuquantum.cutensornet.cutensornet.check_status
cuquantum.cutensornet.cutensornet.cuTensorNetError: CUTENSORNET_STATUS_NOT_SUPPORTED
```

**Comment by yangcal** [comment]
Hello,

Sorry for the delayed response. A couple of notes here:

1. Slicing should be enabled by default so you don't need to specify it in `optimize_options`.

2. `optimize_options` can be provided to either `contract` or `contract_path`. If you pass it to `contract_path`, you'll get a path and the info for the contraction, which can then be passed to `contract` to reuse that path. Or you can directly pass it to `contract` which will perform the two tasks internally. However, if you pass the same `optimize_options` to `contract_path` and `contract` (without `path`), you'll be running path-finding twice. I've attached a snippet below for encouraged usage. 

```python
# Approach I. Reuse Path
optimize_options = {"samples": 5, ...} 
path, info = contract_path(expr, *operands, optimize=optimize)
output = contract(expr, *operands, optimize={'path': path})

# Approach 2. Direct
output = contract(expr, *operands, optimize=optimize)
```

3. The execution failed at path-finding stage before actual contraction is performed. It feels like the path finder is not happy with the optimize_options. Can you turn on CUTENSORNET logging and share the full error message, eg, `CUTENSORNET_LOG_LEVEL=6 python cutn_notebook_benchmark_simplified.py`? Or it would be easiest for us if you can provide a minimal reproducer for this problem. For instance, you could just pickle the contraction expression and operands shapes like below, we'll be able to provide more concrete suggestions on the path-finder setting. 

```python
import pickle
data = {'expr': expr,
        'shapes': [o.shape for o in operands]}
with open('data.pickle', 'wb') as f:
    pickle.dump(data, f)
```

**Reply by rht** [reply]
I have published the code in https://github.com/rht/tensor_network_benchmark. It is runnable with `python main.py`, which by default, will do `contract_path` then `contract` of the https://arxiv.org/abs/2012.02430 QAOA ansatz. The `CUTENSORNET_LOG_LEVEL=6` is too verbose for me to include in here. If that repo is not portable enough, here is the circuit specification: https://github.com/rht/tensor_network_benchmark/blob/fba5ff33c55f36f8bef214bc1266510d9a3b7b0c/main.py#L33-L58.

This is how I specified the optimize options: https://github.com/rht/tensor_network_benchmark/blob/fba5ff33c55f36f8bef214bc1266510d9a3b7b0c/common_tn.py#L79-L88, and so, that the path finding is done twice should be ruled out.

**Reply by yangcal** [reply]
Thanks for sharing the script. Upon inspection of the logging, we noticed that the error is due to the number of elements of the largest intermediate tensor for that particular path exceeding int32_t limit. Such contraction path is currently not supported though we may be able to tune `optimize_options` to more consistently find a path satisfying that constraint. We will get back to you with more updates on this. 

On a separate note, I noticed a few things in your sample:
1. In `common_tn.py`, you created a cutensornet handle via `cutn.create()`. The handle should be destroyed before exiting the program to avoid memory leak.
2. You can simplify this [line](https://github.com/rht/tensor_network_benchmark/blob/fba5ff33c55f36f8bef214bc1266510d9a3b7b0c/common_tn.py#L86C53-L86C53) with just `optimize={"path": path}`
3. When you create a random graph using `nx.random_regular_graph`, you can set the seed for better reproducibility.

**Reply by yangcal** [reply, marked as answer]
Upon further investigation, we realized that the number of elements exceeding int32_t was due to `slicing` information not passed to `optimize` so that `contract` call did not slice at all. Two simple fixes:

1. Replace `optimize={"path": path}` with `optimize={'path': path, 'slicing': info.slices}` 
2. Instead of a two-step `contract_path` and `contract` call, you can simply do `contract(..., optimize=optimize_options)`. We internally pass both path and slicing.

Let us know if all the issues are resolved with the fixes. BTW, I think with the fix you won't need the custom `optimize_options`.

**Reply by rht** [reply]
Thank you! Both method fixed it, and I have tried several times without fixing the seed. This also means I have to redo all my existing `cutn` benchmarks in the light of the `"slicing": info.slices` fix.

**Reply by rht** [reply]
I read the [`contract_path`](https://github.com/NVIDIA/cuQuantum/blob/c202ee7ad23ce9a5072a32b45d36e8310a20eeb5/python/cuquantum/cutensornet/tensor_network.py#L1032-L1037) and [`contract`](https://github.com/NVIDIA/cuQuantum/blob/c202ee7ad23ce9a5072a32b45d36e8310a20eeb5/python/cuquantum/cutensornet/tensor_network.py#L966-L979). From this, I'm not 100% sure that the path and slices are sufficient information from `contract_path` to pass to a `contract` function. This is because the `Network` object may hold other necessary information.

The use case is that I want to to `contract_path` first to know how expensive the contraction operation is going to be, and want to reuse the path to feed it into a `contract` function.

It seems safer for me to do this instead:
```python
def contract_path_return_network(...):
    network = Network(...)
    opt_info = network.contract_path(optimize=optimize)
    return network, opt_info
```

Later on, I analyze `opt_info`, and if the path is satifactory, then I do `output = network.contract(stream=stream)`.

**Comment by rht** [comment]
After the fix in https://github.com/NVIDIA/cuQuantum/discussions/63#discussioncomment-6388650 and upgrading to 23.06, I occasionally encountered this error, with the same circuit I have been using:

```
$ python main.py 
22
QAOA ansatz with cost included
Running with cutn
Peak memory usage: 4.62 gigabytes
Elapsed cutn 0.584
Output cutn (1.3877787807814457e-17+2.1250362580715887e-17j)

24
QAOA ansatz with cost included
Running with cutn
Peak memory usage: 8.74 gigabytes
Elapsed cutn 0.916
Output cutn (-7.806255641895632e-18+1.0842021724855044e-19j)

30
QAOA ansatz with cost included
Running with cutn
Traceback (most recent call last):
  File "main.py", line 124, in <module>
    run_exp("alexeev")
  File "main.py", line 103, in run_exp
    output = run_multiple_methods(
  File "main.py", line 49, in run_multiple_methods
    out, elapsed = common_tn.run_with_cutn(circuit, pauli_string)
  File "/home/cuquantum/tensor_network_benchmark/common_tn.py", line 103, in run_with_cutn
    output = contract(
  File "/home/cuquantum/conda/envs/cuquantum-23.03/lib/python3.8/site-packages/cuquantum/cutensornet/tensor_network.py", line 968, in contract
    output = network.contract(stream=stream)
  File "/home/cuquantum/conda/envs/cuquantum-23.03/lib/python3.8/site-packages/cuquantum/cutensornet/_internal/utils.py", line 474, in inner
    result = wrapped_function(*args, **kwargs)
  File "/home/cuquantum/conda/envs/cuquantum-23.03/lib/python3.8/site-packages/cuquantum/cutensornet/_internal/utils.py", line 474, in inner
    result = wrapped_function(*args, **kwargs)
  File "/home/cuquantum/conda/envs/cuquantum-23.03/lib/python3.8/site-packages/cuquantum/cutensornet/_internal/utils.py", line 474, in inner
    result = wrapped_function(*args, **kwargs)
  File "/home/cuquantum/conda/envs/cuquantum-23.03/lib/python3.8/site-packages/cuquantum/cutensornet/tensor_network.py", line 726, in contract
    self._allocate_workspace_memory_perhaps(stream, stream_ctx)
  File "/home/cuquantum/conda/envs/cuquantum-23.03/lib/python3.8/site-packages/cuquantum/cutensornet/_internal/utils.py", line 474, in inner
    result = wrapped_function(*args, **kwargs)
  File "/home/cuquantum/conda/envs/cuquantum-23.03/lib/python3.8/site-packages/cuquantum/cutensornet/_internal/utils.py", line 474, in inner
    result = wrapped_function(*args, **kwargs)
  File "/home/cuquantum/conda/envs/cuquantum-23.03/lib/python3.8/site-packages/cuquantum/cutensornet/_internal/utils.py", line 443, in inner
    raise e
  File "/home/cuquantum/conda/envs/cuquantum-23.03/lib/python3.8/site-packages/cuquantum/cutensornet/_internal/utils.py", line 435, in inner
    result = wrapped_function(*args, **kwargs)
  File "/home/cuquantum/conda/envs/cuquantum-23.03/lib/python3.8/site-packages/cuquantum/cutensornet/tensor_network.py", line 362, in _allocate_workspace_memory_perhaps
    self.workspace_ptr = self.allocator.memalloc(self.workspace_size)
  File "/home/cuquantum/conda/envs/cuquantum-23.03/lib/python3.8/site-packages/cuquantum/cutensornet/memory.py", line 128, in memalloc
    cp_mem_ptr = cp.cuda.alloc(size)
  File "cupy/cuda/memory.pyx", line 730, in cupy.cuda.memory.alloc
  File "cupy/cuda/memory.pyx", line 742, in cupy.cuda.memory.alloc
  File "cupy/cuda/memory.pyx", line 1419, in cupy.cuda.memory.MemoryPool.malloc
  File "cupy/cuda/memory.pyx", line 1440, in cupy.cuda.memory.MemoryPool.malloc
  File "cupy/cuda/memory.pyx", line 1120, in cupy.cuda.memory.SingleDeviceMemoryPool.malloc
  File "cupy/cuda/memory.pyx", line 1141, in cupy.cuda.memory.SingleDeviceMemoryPool._malloc
  File "cupy/cuda/memory.pyx", line 1379, in cupy.cuda.memory.SingleDeviceMemoryPool._try_malloc
cupy.cuda.memory.OutOfMemoryError: Out of memory allocating 13,548,336,128 bytes (allocated so far: 4,311,993,856 bytes).
```
The error message is definitely different this time. But I wonder because this is due to me not gc-ing the `cupy` handle's memory?

**Reply by yangcal** [reply]
Hello, just to confirm, you're still using the 40GB A100 right? The error message suggests that device memory allocation for 13GB failed (4GB already allocated). That amounts to 17GB and should totally fit into the 40GB A100. One potential caveat is, are you the sole user of the workstation? If another GPU job gets launched around the same time, it might be taking away some device space and lead to the error. 

BTW, can you try the following little script to make sure that your GPU is working as expected?
```python
import cupy as cp
a = cp.ones(4*1024**3,)
```

**Reply by rht** [reply]
It failed until `1*1024**3` works. Upon checking the hardware, it's no longer 40 GB A100. I had recently moved to AWS and had assumed the hardware is A100, but it is actually `Tesla V100-SXM2-16GB`. This solves the mystery! Does this mean there is a limit to the slicing method in reducing the memory? Is it possible to trade for more memory saving in return for longer runtime?

**Reply by yangcal** [reply]
In general, there is no limit on the slicing method to reducing the memory eg, If you slice all modes, it can always fit into GPU :) .

In your current github script, I see you have the memory_limit as 80%, which means the largest allowed memory for cutensornet is your **total GPU memory** * 80%. cuquantum will try to find a path using slicing such that the contraction fits under that constraint. In your case cuquantum found a path that requires ~13GB (that feels slightly larger than 16*80%, can you again verify your device?), but at that time, you've already have allocated 4GB that cupy can't reuse (maybe for other tasks or are there any other users occupying the gpu memory?) 

One way to adjust to gpu mem at runtime is to query the available GPU mem when setting `memory_limit`. You can query the available mem, eg, with `mem_avail = cupy.cuda.Device().mem_info[0]` and then multiply by say 80%. However, if there are other users on the system and they launch gpu job after this query, mem allocation may still fail. 

Meanwhile, I would encourage you to track the baseline gpu usage before the contraction call. For instance, you can simply print out the available memory using the command above before contraction. That may help track down where the rest occupied memory was spent on (some of them will be spent on the input operands for sure, what about the others?).

**Reply by rht** [reply]
I am the sole user of the instance, and there is no other running process in the background.

The only thing that may take the available memory is `opt_einsum`, `cusv`, or Qiskit MPS that I run sequentially in a loop, but that they don't necessarily free the allocated GPU memory.

I suppose `mem_avail = cupy.cuda.Device().mem_info[0]` is the surefire way to do it. I will test this later today.

**Reply by rht** [reply]
I had to explicitly delete the `cutn` output, and call `mempool.free_all_blocks()` right after (https://github.com/rht/tensor_network_benchmark/blob/70d2bc884eeb8fc294983dc9244c28dec059685d/common_tn.py#L284-L286) in order for the available memory to be always 16 GB. The output is already being out of scope, but still it doesn't get gc-ed.

**Reply by rht** [reply]
> The output is already being out of scope, but still it doesn't get gc-ed.

Actually, the running in various frameworks (`cutn`, `opt_einsum`, etc) happens within the same scope, so it makes sense that the tensor output really has to be explicitly deleted.

**Reply by yangcal** [reply]
cupy has a reuse mechanism and does not immediately release the allocated memory that goes out of scope. That's why the mem usage will still be non-zero even though the contraction goes out of scope. 

Can you turn on logging in next few experiments? Not CUTENSORNET_LOG_LEVEL, but with
```python
import logging
logging.basicConfig(level=logging.INFO, format='%(asctime)s %(levelname)-8s %(message)s', datefmt='%m-%d %H:%M:%S')
```

In my experiments where I only tested cutn executions, I've consistently observed a baseline usage of 0.9GB (probably for operands storage) before the contract call in each iteration. I'm trying to figure out where your 4GB usage is from (cupy determines that these 4GB can't be reused or freed).

**Reply by rht** [reply]
Sorry for the late reply. Here is the stdout and stderr output: https://gist.github.com/rht/e67ab461b24255653398676ccb2d5345. This is with `memory_limit` set to `0.8 * available_memory`, which understandably gives 12 GB. But even after commenting out https://github.com/rht/tensor_network_benchmark/blob/3b3dd4b14e7cfda9bf43448401c5eb121709a963/common_tn.py#L72, I still get 12 GB.

According to https://docs.nvidia.com/cuda/cuquantum/latest/python/api/generated/cuquantum.NetworkOptions.html#cuquantum.NetworkOptions.memory_limit, 80% is the default?

**Reply by yangcal** [reply]
The 80% default indeed refers to the total mem * 0.8. With `available_memory *.8`, it shall be lower than the default based on the amount of memory that has been occupied. If you're seeing close numbers for both, that means the resource are being freed in time and the only difference between the two would be the small baseline usage (for instance, your input cupy operands). In an ideal case, your baseline usage should be low (and that's now reflected in your latest run). You can always print out both `mem_info[0]`(available memory) and `mem_info[1]` (total memory) to find out your baseline usage and see if that's reasonable. Anyway, your script now monitors the free mem at runtime and shall always succeed regardless of baseline usage (though I'm still curious on where your previous 4GB baseline usage was from). 

BTW, your conversion from bytes to GB is not correct. The 16GB here equals 16*1024**3 bytes for our hardware.

**Reply by rht** [reply]
Thank you, I have fixed the conversion factor.

> (though I'm still curious on where your previous 4GB baseline usage was from)

I am able to consistently reproduce the OOM at this commit: https://github.com/rht/tensor_network_benchmark/commit/c129894f92839408b1638f8f545b59a4cd21c735 (with MPS disabled, because MPS can't run on that QAOA ansatz). This is without deleting the computation output that still lives in the GPU, so it makes sense there is a leftover 4 GB from a previous computation (via `opt_einsum`) within the same scope.

> cupy has a reuse mechanism and does not immediately release the allocated memory that goes out of scope. 

I tested. Deleting the `cutn` and `opt_einsum` output on their own is not sufficient. My guess is that `mempool.free_all_blocks()` syncs the Python process state with GPU that the output is actually deleted.

**Reply by yangcal** [reply]
Just to confirm, you're seeing OOM when both `opt_einsum` and `cutn` backend are called?

It seems you're computing the expectation value as output and therefore the output itself should be tiny? My guess with `opt_einsum` is that it dispatches to `cupy` for pairwise contractions which creates some intermediate tensors that are cached in mempool. The intermediate tensors may not be freed even after it goes out of scope, unless `mempool.free_all_blocks()` is called explicitly right after. 

Note that deleting output/intermediate or them going out of scope (both meaning reference count dropping to 0) will not trigger cupy to free that cache. The weird part is cupy should be able to reuse that 4GB cache if the reference count is really 0 with process synced (no explicit "free_all_blocks()" call needed).

**Reply by rht** [reply]
> Just to confirm, you're seeing OOM when both opt_einsum and cutn backend are called?

Yes. I tested that I didn't run into OOM if I did `cutn` several times in the same scope. But this could be because `cutn` has automatic slicing (that it uses less memory), and `opt_einsum` doesn't.
